### PR TITLE
:sparkles: Add flag to allow use best effort resource for klusterlet

### PR DESCRIPTION
**Pull request by haowells**
## Summary
see PR in https://github.com/open-cluster-management-io/ocm/pull/317

## Related issue(s)

Fixes #

### CODE REVIEW

**Comment by qiujian16** [comment]
would need to run `make update` to update deepcopy and client code

File: operator/v1/types_klusterlet.go [hunk 1]
```diff
@@ -86,6 +86,10 @@ type KlusterletSpec struct {
 	// registration-agent and work-agent will use it to communicate with hub api server.
 	// +optional
 	HubApiServerHostAlias *HubApiServerHostAlias `json:"hubApiServerHostAlias,omitempty"`
+
+	// BestEffortResource allow use the best effort resource for operator/registration/work deployment
+	// +optional
+	BestEffortResource bool `json:"bestEffortResource,omitempty"`
```
Comment (qiujian16) [inline operator/v1/types_klusterlet.go]: It is better no to use boolean in API. You might consider a resources struct with a type in it. We can just set type to "BestEffort" and "Default" today, and add other option to be able to specify resource in the future.

Comment (haowells) [reply operator/v1/types_klusterlet.go]: @qiujian16  thanks. address your comments and PTAL again.

File: operator/v1/types_klusterlet.go [hunk 2]
```diff
@@ -86,6 +86,22 @@ type KlusterletSpec struct {
...
+type ResourceQosClass string
+
+const (
+	ResourceQosClassDefault    ResourceQosClass = "Default"
+	ResourceQosClassBestEffort ResourceQosClass = "BestEffort"
+)
+
+// ResourceType allow user override the default pod QoS classes
+type ResourceType struct {
+	Type ResourceQosClass `json:"type"`
```
Comment (qiujian16) [inline operator/v1/types_klusterlet.go:108]: // +kubebuilder:validation:Enum=Default;BestEffort
  // +kubebuilder:default:=Default

Should add some valildation

File: operator/v1/types_klusterlet.go [hunk 3]
```diff
@@ -86,6 +86,22 @@ type KlusterletSpec struct {
 	// registration-agent and work-agent will use it to communicate with hub api server.
 	// +optional
 	HubApiServerHostAlias *HubApiServerHostAlias `json:"hubApiServerHostAlias,omitempty"`
+
+	// ResourceType specify QoS classes of klusterlet deployment
+	// +optional
+	ResourceType ResourceType `json:"resourceType,omitempty"`
```
Comment (qiujian16) [inline operator/v1/types_klusterlet.go]: Not sure ResourceType is a good name. How about `ResourceRequest` or `ResourceRequirement`?

Comment (haowells) [reply operator/v1/types_klusterlet.go]: ok. let's use ResourceRequirement

File: operator/v1/types_klusterlet.go [hunk 4]
```diff
@@ -86,6 +86,24 @@ type KlusterletSpec struct {
...
+	// ResourceRequirement specify QoS classes of klusterlet deployment
+	// +optional
+	ResourceRequirement *ResourceRequirement `json:"resourceRequirement,omitempty"`
+}
+
+type ResourceQosClass string
+
+const (
+	ResourceQosClassDefault    ResourceQosClass = "Default"
+	ResourceQosClassBestEffort ResourceQosClass = "BestEffort"
```
Comment (qiujian16) [inline operator/v1/types_klusterlet.go:101]: some comments maybe to indicate what besteffort means

Comment (haowells) [reply operator/v1/types_klusterlet.go:101]: comments added

**Comment by qiujian16** [comment]
need to signoff. And you would need to run make update again.

**Pull request merged**